### ISSUE TITLE
AnyUrl adds trailing slash

### ISSUE THREAD

**Issue by charel-felten-rq** [pydantic/pydantic#7186]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

AnyURL now adds a trailing slash in validation of URLs that dont have a trailing slash. This caused CORS issues for us, because CORS doesnt allow trailing slash (I think, seems to be written here: https://www.w3.org/TR/2008/WD-access-control-20080214/)

In my opinion i see this as a bug because I dont see the option toggle this addition of trailing slash. Maybe its expected behaviour, please let me know then how I can turn it off.

### Example Code

```Python
from pydantic import AnyHttpUrl as AnyUrlv2
from pydantic.v1 import AnyHttpUrl as AnyUrlv1

print(AnyUrlv2("http://localhost:3000"))
print(AnyUrlv1("http://localhost:3000", scheme="http"))


prints:

http://localhost:3000/
http://localhost:3000
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.2.1
        pydantic-core version: 2.6.1
          pydantic-core build: profile=release pgo=false
                 install path: /Users/charelfeltenrq/.pyenv/versions/3.10.8/lib/python3.10/site-packages/pydantic
               python version: 3.10.8 (main, Dec  2 2022, 11:54:08) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.3.1-arm64-arm-64bit
     optional deps. installed: ['email-validator', 'typing-extensions']
```

**Comment by charel-felten-rq** [comment]
may be related: https://github.com/pydantic/pydantic-settings/issues/104

**Comment by charel-felten-rq** [comment]
This problem also becomes apparent in the url builder: 

```python
from pydantic import AnyHttpUrl as AnyHttpUrl

AnyHttpUrl.build(
    scheme="https",
    host="google.com",
    path="/search",
)
```

produces: `Url('https://google.com//search')`

Is that the desired outcome? If yes, then does it mean that all Url builds we have in our codebase should now have a path that starts without a slash?

**Comment by samuelcolvin** [comment]
The second case is unrelated.

This is done by the URL rust crate, so it probably mandated by a rfc somewhere.

**Comment by samuelcolvin** [comment]
Sorry, the second case is related.

**Comment by charel-felten-rq** [comment]
Its weird.. so assuming, AnyUrl 'validates' paths by ensuring that they are conform as they end with a trailing slash, then why does it insert a second slash here again?

```python
from pydantic import AnyHttpUrl as AnyHttpUrl

AnyHttpUrl.build(
    scheme="https",
    host="google.com/",
    path="search",
)
```

`Url('https://google.com//search')`

**Comment by adriangb** [comment]
> Another weird situation

That's technically valid and works, not sure we need to do anything special there.

I think we might come up with a better solution for this long term but for now if you don't need a `Url` instance you could do something like:


```python
from typing import Annotated, Any, List

from pydantic_core import CoreSchema, core_schema

from pydantic import AfterValidator, BeforeValidator, GetCoreSchemaHandler, TypeAdapter
from pydantic.networks import AnyUrl


class Chain:
    def __init__(self, validations: List[Any]) -> None:
        self.validations = validations
    
    def __get_pydantic_core_schema__(self, source_type: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
        return core_schema.chain_schema(
            [
                *(handler.generate_schema(v) for v in self.validations),
                handler(source_type)
            ]
        )


def strip_slash(value: str) -> str:
    return value.rstrip('/')


MyAnyUrl = Annotated[str, AfterValidator(strip_slash), BeforeValidator(str), Chain([AnyUrl])]

ta = TypeAdapter(MyAnyUrl)

print(ta.validate_python('https://www.google.com'))
# https://www.google.com
```

**Comment by charel-felten-rq** [comment]
Thank you for posting this workaround, I am not 100% sure whats going on there, but to my understanding it seems that it creates a type `MyAnyUrl` which is a str, but includes the validation from `AnyUrl`. That could be useful, maybe we will use it or a similar idea.

However, I am still a bit baffled that it seems to be seen as not a problem that AnyUrl adds trailing `\` to all Urls, to me this is not validation anymore as it actually changes the Url. And also `.build` creating Urls with `\\` in them is problematic for our situation. 

For now, we have decided to roll back to pydantic v1. Will there be any way to keep using the pydantic v1 Url behaviour (i.e. not adding `\`) in v2? We tried via `from pydantic.v1 import AnyUrl` but this caused errors inside `BaseSettings`. Or is this really the desired behaviour and you are not planning on changing it back?

Lastly, just wanted to mention, I am a very happpy user of pydantic, and I love the work you all do, just this change of behaviour of Url is giving us a bit of a headache.

**Comment by tworedz** [comment]
Also finding this new behavior very annoying. In pydantic v1 urls inherited from str, so it can be easily passed to library codes, where they usually expects string representation of urls. But now we need to `str(model.url_field)` or `model.url_field.unicode_string()` when passing urls, because now it is `pydantic_core._pydantic_core.Url`

I'm not sure which variant right (old or new url), but this little slash causes very strange and hours of uncatchable bug fixing.

FYI, I come up with this solution:
```python
from typing import Annotated
from pydantic import AfterValidator, AnyUrl, BaseModel

URL = Annotated[AnyUrl, AfterValidator(lambda x: str(x).rstrip("/"))]

class Model(BaseModel):
    url: URL

print(Model(url="https://google.com").url)
# > https://google.com
```
But sadly we cannot use this solution everywhere, because sometimes we need the last slash.

Seems, that v2 url validator adding slashes only to `naked` urls, i.e. without path, which is acceptable:
```python
from pydantic import AnyUrl, BaseModel

class Model(BaseModel):
    url: AnyUrl

print(Model(url="https://google.com").url)
# > https://google.com/

print(Model(url="https://google.com/").url)
# > https://google.com/

print(Model(url="https://google.com/api").url)
# > https://google.com/api

print(Model(url="https://google.com/api/").url)
# > https://google.com/api/
```

Would be nice, if old behavior preserves, or like author suggested, we can choose behavior type. Thanks in advance!

**Comment by Tradunsky** [comment]
+1 this is a bug.

In addition to a broken behavior when config URLs used to be used to append a path, like:
https://google.com +"/search" (and all other variants with appending path in python)

There is also a violated principle of a least surprise when original value modified by a validation framework as when an URL is configured without slash is returned or saved in other components: DB, external services, they receive a new value which silently violated previous protocol until those actually start to execute it in a request.

**Comment by chbndrhnns** [comment]
> That's technically valid and works

@adriangb 
It might be technically valid. However, a URL containing a double slash is pointing to a different endpoint (not sure about the semantic) and thus I expect  webservers to respond with a 404.

**Comment by alexsantos** [comment]
I had some validation checks because of this issue: the validator was testing for `http://loinc.org` but I was getting `http://loinc.org/` on the `model_dump_json()`
This is what I did:

`from typing import Optional

from pydantic import BaseModel, AnyUrl, Extra, field_serializer


class Coding(BaseModel):
    system: Optional[AnyUrl] = None
    version: Optional[str] = None
    code: Optional[str] = None
    display: Optional[str] = None
    userSelected: Optional[bool] = None

    @field_serializer('system')
    def serialize_dt(self, system: AnyUrl):
        return str(system).rstrip('/')

    class Config:
        extra = Extra.forbid`

**Comment by Saran33** [comment]
Another less-than-ideal workaround (if it's acceptable to strip all trailing slashes):

```python
from typing import Annotated
from pydantic import AnyHttpUrl, AfterValidator, BaseModel, PlainValidator, TypeAdapter

AnyHttpUrlAdapter = TypeAdapter(AnyHttpUrl)

HttpUrlStr = Annotated[
    str,
    PlainValidator(lambda x: AnyHttpUrlAdapter.validate_strings(x)),
    AfterValidator(lambda x: str(x).rstrip("/")),
]


class Model(BaseModel):
    url: HttpUrl


url = Model(url="https://google.com").url
print(url)
# > https://google.com
```

In addition to the above points, IMO, when appending paths, this is more readable:
```python
URL = f"{BASE_URL}/api/v1"
```
than this:
```python
URL = f"{BASE_URL}api/v1"
```

**Comment by rafrafek** [comment]
I tried:

```python
service_url: HttpUrl

@field_validator("service_url", mode="after")
@classmethod
def service_url_remove_trailing_slash(cls, service_url: HttpUrl) -> str:
    return str(service_url).rstrip("/")
```

But then I got warnings:
```
/site-packages/pydantic/main.py:308: UserWarning: Pydantic serializer warnings:
    Expected `url` but got `str` - serialized value may not be as expected
    return self.__pydantic_serializer__.to_python(
```

How to keep the field as `HttpUrl` without the trailing slash?

**Comment by slingshotvfx** [comment]
> Seems, that v2 url validator adding slashes only to `naked` urls, i.e. without path, which is acceptable:
> 
> ```python
> from pydantic import AnyUrl, BaseModel
> 
> class Model(BaseModel):
>     url: AnyUrl
> 
> print(Model(url="https://google.com").url)
> # > https://google.com/
> 
> print(Model(url="https://google.com/").url)
> # > https://google.com/
> 
> print(Model(url="https://google.com/api").url)
> # > https://google.com/api
> 
> print(Model(url="https://google.com/api/").url)
> # > https://google.com/api/
> ```

This inconsistency is my biggest issue, especially when trying to add on paths or sub-paths as others have mentioned. If you can't be certain if a url has a trailing slash or not out of the box then we almost need an `os.path.join` style method that ensures exactly one forward slash between each path component.

**Comment by pavradev** [comment]
We got this problem when we migrated from pydantic 1 as well. In addition to that, a lot of code was broken since AnyUrl is not accepted as str any more. This is our solution if it will be helpful to anyone:
`
AnyUrlTypeAdapter = TypeAdapter(AnyUrl)
AnyUrlStr = Annotated[
    str,
    BeforeValidator(lambda value: AnyUrlTypeAdapter.validate_python(value) and value),
]
`
So we currently just use our "wrapper" AnyUrlStr everywhere for compatibility. It will also keep your original string without any modification if it passes the validation.

**Comment by JensHeinrich** [comment]
Wouldn't it be better to enhance the `AnyUrl` type by giving it it's own implementation of `__add__` providing a feature like `urllib`'s [`urljoin`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin) ?

**Comment by vgavro** [comment]
It's **definitely** a BUG, not "documentation issue".
Consider this example:
```
In [14]: url = AnyUrl.build(scheme='http', host='example.com', path='/')

In [15]: url = AnyUrl.build(scheme=url.scheme, host=url.host, path=url.path)

In [16]: url = AnyUrl.build(scheme=url.scheme, host=url.host, path=url.path)

In [17]: print(url)
http://example.com////
```

**Comment by samuelcolvin** [comment]
Thanks everyone for commenting.

I'd be happy to accept a PR to allow an opt-in change of behaviour. While i know many people here think the current behaviour is a bug, there will be others who are relying on it, so we can't change the default behavior in a minor release.

Here's my proposal for what needs to change:
* change [this code ](https://github.com/pydantic/pydantic-core/blob/758bc51d4bb7e94aae0c8e6f4ce515b2d55cc3a6/src/url.rs#L23) to add another fiels to the `PyUrl` rust struct called something like `omit_trailing_slash: bool`, default `false` only applies when `lib_url.path == "/"`
* Based on that flag, change the behaviour of `__str__`, `__repr__`, `unicode_string` and `path` to omit the trailing slash if set
* Same changes for [`PyMultiHostUrl`](https://github.com/pydantic/pydantic-core/blob/758bc51d4bb7e94aae0c8e6f4ce515b2d55cc3a6/src/url.rs#L219C6-L219C20)
* change the [`url_schema()`](https://github.com/pydantic/pydantic-core/blob/758bc51d4bb7e94aae0c8e6f4ce515b2d55cc3a6/python/pydantic_core/core_schema.py#L3476) core schema to accept `omit_trailing_slash`
* same for `multi_host_url_schema`
* we'll then need a PR to pydantic to add a config flag `url_omit_trailing_slash` (default `False`) which changes how the core schema is built

If anyone's interested, feel free to submit a PR. Might be a nice start for anyone interested in getting their hands dirty with rust as (AFAIK) it shouldn't be a too complicated change, but also not trivial.

**Comment by tworedz** [comment]
If no one is against it, I want to try!

**Comment by samuelcolvin** [comment]
@tworedz, go for it.

> Wouldn't it be better to enhance the `AnyUrl` type by giving it it's own implementation of `__add__` providing a feature like `urllib`'s [`urljoin`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin) ?

@JensHeinrich, interesting idea - PR welcome.

**Comment by vgavro** [comment]
@samuelcolvin that's the problem with this ticket - it puts the problem wrong.

The problem is not with **trailing** slashes, it's about **leading** path slash.

`/api` and `/api/` are different paths, pydantic handles them correctly, flag name `omit_trailing_slash` would be confusing!

There are TWO problems:

1. Behavior with **normalizing** urls  - adding slash to path if path is empty. From HTTP protocol point of view it makes perfect sense, (at least HTTP 1.x family, because you can't send a request to "" path) - but for other protocols may not. So option like "ensure_absolute_path=True" is possible, enabled by default for compatibility, and may be overridden, but I believe this behavior perfectly makes sense for HTTP by default.

2. **BUG* with `Url.build` (why I created separate https://github.com/pydantic/pydantic/issues/8640 for that) - that adding LEADING slashes no matter if there WAS leading slash, or not. Strictly saying, it doesn't relate to normalizing urls problem.

**Comment by davidhewitt** [comment]
For what it's worth, the outcome of https://github.com/pydantic/pydantic-core/pull/1173 was that we think it's best to get the following upstream feature added for us to do `Url.build` properly: https://github.com/servo/rust-url/issues/835

Based on the upvotes on that issue, anyone who is interested in taking that on would make a lot of people happy!

**Comment by kfreezen** [comment]
I keep running into this issue in a project I work on. The real problem isn't omission of slashes, it's the fact that hostnames get slashes added to them **even if** the user or developer did not specify them.

So
```
http://localhost:3000 -> http://localhost:3000/
```

But
```
http://localhost:3000/api -> http://localhost:3000/api
```

I feel like I've been on a discussion about this somewhere already, but I keep running into this issue. If we had a way to make it so that we could use the type validation features but not add a slash to hostname (like it was in v1) that would be fantastic.

@samuelcolvin would `omit_trailing_slash: bool` then omit slashes in all cases, even if user specifies them? I don't think it should, because then you have the same problem as before, except now you're just omitting the slash instead of adding it.

**Comment by shawnwall** [comment]
also having this issue with CORS and need to work around

**Comment by bechtold** [comment]
Also ended up here because my FastAPI allow_origins weren't working anymore. 20 minutes debugging, oh well.
I'll just remove the trailing slash for now, with `rstrip('/')`:

```python
[str(origin).rstrip('/') for origin in settings.BACKEND_CORS_ORIGINS]
```

**Comment by patrsc** [comment]
I created a Python package [pydantic-string-url](https://pypi.org/project/pydantic-string-url/), which can be used as a replacement for Pydantic's URL types `AnyUrl`, `AnyHttpUrl`, `HttpUrl`, `AnyWebsocketUrl`, `WebsocketUrl`, `FileUrl`, and `FtpUrl`. Those types are based on `str` and can be directly passed to library packages that expect strings. They also preserve the absence of trailing slashes. Usage example:

```py
from pydantic import BaseModel, TypeAdapter, ValidationError
from pydantic_string_url import HttpUrl

class User(BaseModel):
    name: str
    age: int
    homepage: HttpUrl

user = {"name": "John Doe", "age": 33, "homepage": "https://example.com"}
invalid_user = {"name": "Alice", "age": 32, "homepage": "not a url"}

john = User.model_validate(user)
assert john.homepage == "https://example.com"  # no trailing slash was added
try:
    alice = User.model_validate(invalid_user)
except ValidationError as e:
    print(str(e))
```

[Feedback](https://github.com/patrsc/pydantic-string-url/issues) and bug reports are welcome. Maybe something similar makes it into Pydantic in the future.

**Comment by xuanlinh91** [comment]
Still have this bug in Oct 2024, please fix

**Comment by MarcBresson** [comment]
This bug still exists in 2.10.4

**Comment by sydney-runkle** [comment]
Indeed, I don't think we'll make many url related changes until v3, the minor ones we made in v2.10 resulted in an abundance of patch fixes needing to be made :(

**Comment by shurshilov** [comment]
The validation library, instead of simply validating, modifies the data by adding a '/' at the end...damn, what a low-quality code in pidantics, I do not advise using it in new projects.

**Comment by SnoozeFreddo** [comment]
+1....

**Comment by HenriBlacksmith** [comment]
I agree that there should be a mean to preserve or access the original formatting coming from the user input when it is valid.

In some cases like OAuth 2.0 redirect URIs, adding a trailing slash can cause issues as validating a redirect URI MUST be an exact match. 

I am building systems validating redirect URIs and I am facing the issue of the trailing slash being added and causing the redirect URI to be rejected when it does not have one.

I am not exactly sure to understand the downvotes on the post above: https://github.com/pydantic/pydantic/issues/7186#issuecomment-2678492196

It makes sense to preserve the user input when it is valid

**Comment by samuelcolvin** [comment]
Sorry, we should have fixed this long ago, sorry we dropped the ball.

I have a PR up on pydantic-core to add a config switch to avoid adding trailing slashes, https://github.com/pydantic/pydantic-core/pull/1719.

Unfortunately I don't think we can change the default until V3.

**Comment by Viicos** [comment]
Forgot to mention it here, but the 2.12 release shipped with a new option to remove trailing slash (see the [blog post entry](https://pydantic.dev/articles/pydantic-v2-12-release#preserve-empty-url-paths)).
